ostd/src/mm/frame/options.rs

```diff
@@ -23,111 +23,62 @@ use crate::{
+    /// improve performance, but it definitely introduces security
+    /// vulnerabilities by leaking sensitive information. The caller is
+    /// responsible for writing to all the bits of the allocated frames
+    /// before sharing them with other components.
+    pub fn zero_init(&mut self, zero_init: bool) -> &mut Self {
+        self.zero_init = zero_init;
         self
     }
 
-    /// Allocates a collection of page frames according to the given options and metadata.
-    ///
-    /// The metadata function `metadata_fn` is called for each allocated frame to provide metadata.
-    pub fn alloc<F, M>(&self, mut metadata_fn: F) -> Result<Vec<Frame<M>>>
-    where
-        F: FnMut(Paddr) -> M,
-        M: FrameMeta,
-    {
-        let pages = if self.is_contiguous {
-            page::allocator::alloc(self.nframes * PAGE_SIZE, |p| {
-                FrameMetaBox::new(metadata_fn(p))
-            })
-            .ok_or(Error::NoMemory)?
-        } else {
-            page::allocator::alloc_contiguous(self.nframes * PAGE_SIZE, |p| {
-                FrameMetaBox::new(metadata_fn(p))
-            })
-            .ok_or(Error::NoMemory)?
-            .into()
-        };
-        let frames: Vec<_> = pages
-            .into_iter()
-            .map(|page|
-                // SAFETY: The pages are allocated with the metadata type `M`.
-                unsafe { Frame::<M>::from_unchecked(page.into()) })
-            .collect();
-        if !self.uninit {
-            for frame in frames.iter() {
-                frame.writer().fill(0);
-            }
-        }
-
-        Ok(frames)
-    }
-
-    /// Allocates a single page frame according to the given options and metadata.
+    /// Allocates a single page frame with provided metadata.
     pub fn alloc_single<M: FrameMeta>(&self, metadata: M) -> Result<Frame<M>> {
-        if self.nframes != 1 {
-            return Err(Error::InvalidArgs);
-        }
-
         let page =
             page::allocator::alloc_single(FrameMetaBox::new(metadata)).ok_or(Error::NoMemory)?;
         // SAFETY: The pages are allocated with the metadata type `M`.
         let frame = unsafe { Frame::<M>::from_unchecked(page.into()) };
-        if !self.uninit {
+        if self.zero_init {
             frame.writer().fill(0);
         }
 
         Ok(frame)
     }
 
-    /// Allocates a contiguous range of page frames according to the given options.
+    /// Allocates a contiguous range of page frames with provided metadata.
     ///
-    /// The returned [`Segment`] contains at least one page frame.
-    pub fn alloc_contiguous<M, F>(&self, mut metadata_fn: F) -> Result<Segment>
+    /// The metadata is initialized with the caller-provided closure
+    /// `metadata_fn`. The closure receives the physical address of the
+    /// page and returns the metadata for the page.
+    pub fn alloc_contiguous<M, F>(&self, nframes: usize, mut metadata_fn: F) -> Result<Segment>
```

Comment:
In the other comment of mine, I suggest renaming `AnyFrame` to `Frame` because we want make the common use cases simple, and advanced use cases possible. In the same spirit, I suggest making the API for allocating frames without user-defined metadata easier. For the simplest use case, the refactored interface is cumbersome to use.

```rust
let single_frame = options.alloc_single(()).unwrap();

let contiguous_segment = options.alloc_contiguous(10, |_| ()).unwrap();
```

Let's say that you have adopted the name `Frame` (currently known as `AnyFrame`), `FrameEx<M>` (currently known as `Frame`). And similarly, `Segment` and `SegmentEx`. Then, the allocation API can be revised to the following form.

```rust
impl FrameAllocOptions {
    pub fn alloc_single(&self) -> Result<Frame> { ... }

    pub fn alloc_single_ex<M>(&self) -> Result<FrameEx<M>> { ... }

    pub fn alloc_contiguous(&self, nframes: usize) -> Result<Segment> { ... }

    pub fn alloc_contiguous_ex<M, F>(&self, nframes: usize, mut metadata_fn: F) -> Result<SegmentEx<M>> { ... }
}
```